histogram observer: ensure buffer shape consistency

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44956 histogram observer: ensure buffer shape consistency**

Summary:

Makes buffer shapes for HistogramObserver have the
same shapes in uninitialized versus initialized states.

This is useful because the detectron2 checkpointer assumes
that these states will stay the same, so it removes the
need for manual hacks around the shapes changing.

Test Plan:

```
python test/test_quantization.py TestObserver.test_histogram_observer_consistent_buffer_shape
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23785382](https://our.internmc.facebook.com/intern/diff/D23785382)

vkuzo committed:
histogram observer: ensure buffer shape consistency

Summary:

Makes buffer shapes for HistogramObserver have the
same shapes in uninitialized versus initialized states.

This is useful because the detectron2 checkpointer assumes
that these states will stay the same, so it removes the
need for manual hacks around the shapes changing.

Test Plan:

```
python test/test_quantization.py TestObserver.test_histogram_observer_consistent_buffer_shape
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented on torch/quantization/observer.py:
```diff
@@ -739,8 +739,8 @@ def __init__(self, bins=2048, upsample_rate=128, dtype=torch.quint8,
-        self.register_buffer('min_val', torch.tensor([]))
-        self.register_buffer('max_val', torch.tensor([]))
+        self.register_buffer('min_val', torch.tensor(float('inf')))
+        self.register_buffer('max_val', torch.tensor(float('-inf')))
```
would we need similar changes for PerChannel Observers too? They have the same issue as HistogramObserver

vkuzo committed:
Update on "histogram observer: ensure buffer shape consistency"


Summary:

Makes buffer shapes for HistogramObserver have the
same shapes in uninitialized versus initialized states.

This is useful because the detectron2 checkpointer assumes
that these states will stay the same, so it removes the
need for manual hacks around the shapes changing.

Test Plan:

```
python test/test_quantization.py TestObserver.test_histogram_observer_consistent_buffer_shape
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23785382](https://our.internmc.facebook.com/intern/diff/D23785382)

[ghstack-poisoned]
